File: analytics/metrics-platform/src/main/java/org/wikimedia/metrics_platform/config/CurationFilter.java
package org.wikimedia.metrics_platform.config;

import java.time.Instant;
import java.util.function.Predicate;

import org.wikimedia.metrics_platform.config.curation.CollectionCurationRules;
import org.wikimedia.metrics_platform.config.curation.ComparableCurationRules;
import org.wikimedia.metrics_platform.config.curation.CurationRules;
import org.wikimedia.metrics_platform.context.AgentData;
import org.wikimedia.metrics_platform.context.MediawikiData;
import org.wikimedia.metrics_platform.context.PageData;
import org.wikimedia.metrics_platform.context.PerformerData;
import org.wikimedia.metrics_platform.event.EventProcessed;

import com.google.gson.annotations.SerializedName;

public class CurationFilter implements Predicate<EventProcessed> {
    @SerializedName("agent_app_install_id") CurationRules<String> agentAppInstallIdRules;
    @SerializedName("agent_client_platform") CurationRules<String> agentClientPlatformRules;
    @SerializedName("agent_client_platform_family") CurationRules<String> agentClientPlatformFamilyRules;

    @SerializedName("mediawiki_database") CurationRules<String> mediawikiDatabase;

    @SerializedName("page_id") ComparableCurationRules<Integer> pageIdRules;
    @SerializedName("page_namespace_id") ComparableCurationRules<Integer> pageNamespaceIdRules;
    @SerializedName("page_namespace_name") CurationRules<String> pageNamespaceNameRules;
    @SerializedName("page_title") CurationRules<String> pageTitleRules;
    @SerializedName("page_revision_id") ComparableCurationRules<Long> pageRevisionIdRules;
    @SerializedName("page_wikidata_qid") CurationRules<String> pageWikidataQidRules;
    @SerializedName("page_content_language") CurationRules<String> pageContentLanguageRules;

    @SerializedName("performer_id") ComparableCurationRules<Integer> performerIdRules;
    @SerializedName("performer_name") CurationRules<String> performerNameRules;
    @SerializedName("performer_session_id") CurationRules<String> performerSessionIdRules;
    @SerializedName("performer_pageview_id") CurationRules<String> performerPageviewIdRules;
    @SerializedName("performer_groups") CollectionCurationRules<String> performerGroupsRules;
    @SerializedName("performer_is_logged_in") CurationRules<Boolean> performerIsLoggedInRules;
    @SerializedName("performer_is_temp") CurationRules<Boolean> performerIsTempRules;
    @SerializedName("performer_registration_dt") ComparableCurationRules<Instant> performerRegistrationDtRules;
    @SerializedName("performer_language_groups") CurationRules<String> performerLanguageGroupsRules;
    @SerializedName("performer_language_primary") CurationRules<String> performerLanguagePrimaryRules;

    public boolean test(EventProcessed event) {
        return applyAgentRules(event.getAgentData())
                && applyMediaWikiRules(event.getMediawikiData())
                && applyPageRules(event.getPageData())
                && applyPerformerRules(event.getPerformerData());
    }

    private boolean applyAgentRules(@Nonnull AgentData data) {
        return applyPredicate(this.agentAppInstallIdRules, data.getAppInstallId())
                && applyPredicate(this.agentClientPlatformRules, data.getClientPlatform())
                && applyPredicate(this.agentClientPlatformFamilyRules, data.getClientPlatformFamily());
    }

    private boolean applyMediaWikiRules(@Nonnull MediawikiData data) {
        return applyPredicate(this.mediawikiDatabase, data.getDatabase());
    }

    private boolean applyPageRules(@Nonnull PageData data) {
        return applyPredicate(this.pageIdRules, data.getId())
                && applyPredicate(this.pageNamespaceIdRules, data.getNamespaceId())
                && applyPredicate(this.pageNamespaceNameRules, data.getNamespaceName())
                && applyPredicate(this.pageTitleRules, data.getTitle())
                && applyPredicate(this.pageRevisionIdRules, data.getRevisionId())
                && applyPredicate(this.pageWikidataQidRules, data.getWikidataItemQid())
                && applyPredicate(this.pageContentLanguageRules, data.getContentLanguage());
    }

    private boolean applyPerformerRules(@Nonnull PerformerData data) {
        return applyPredicate(this.performerIdRules, data.getId())
                && applyPredicate(this.performerNameRules, data.getName())
                && applyPredicate(this.performerSessionIdRules, data.getSessionId())
                && applyPredicate(this.performerPageviewIdRules, data.getPageviewId())
                && applyPredicate(this.performerGroupsRules, data.getGroups())
                && applyPredicate(this.performerIsLoggedInRules, data.getIsLoggedIn())
                && applyPredicate(this.performerIsTempRules, data.getIsTemp())
                && applyPredicate(this.performerRegistrationDtRules, data.getRegistrationDt())
                && applyPredicate(this.performerLanguageGroupsRules, data.getLanguageGroups())
                && applyPredicate(this.performerLanguagePrimaryRules, data.getLanguagePrimary());
    }

    private static <T> boolean applyPredicate(@Nullable Predicate<T> rules, @Nullable T value) {
        return rules == null || rules.test(value);
    }
}

File: analytics/metrics-platform/src/main/java/org/wikimedia/metrics_platform/config/CurationFilter.kt
@file:UseSerializers(InstantSerializer::class)

package org.wikimedia.metrics_platform.config

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import org.wikimedia.metrics_platform.config.curation.CollectionCurationRules
import org.wikimedia.metrics_platform.config.curation.ComparableCurationRules
import org.wikimedia.metrics_platform.config.curation.CurationRules
import org.wikimedia.metrics_platform.context.AgentData
import org.wikimedia.metrics_platform.context.InstantSerializer
import org.wikimedia.metrics_platform.context.MediawikiData
import org.wikimedia.metrics_platform.context.PageData
import org.wikimedia.metrics_platform.context.PerformerData
import org.wikimedia.metrics_platform.event.EventProcessed
import java.time.Instant
import java.util.function.Predicate

@Serializable
class CurationFilter : Predicate<EventProcessed> {
    @SerialName("agent_app_install_id") var agentAppInstallIdRules: CurationRules<String?>? = null
    @SerialName("agent_client_platform") var agentClientPlatformRules: CurationRules<String?>? = null
    @SerialName("agent_client_platform_family") var agentClientPlatformFamilyRules: CurationRules<String?>? = null
    @SerialName("mediawiki_database") var mediawikiDatabase: CurationRules<String>? = null
    @SerialName("page_id") var pageIdRules: ComparableCurationRules<Int>? = null
    @SerialName("page_namespace_id") var pageNamespaceIdRules: ComparableCurationRules<Int>? = null
    @SerialName("page_namespace_name") var pageNamespaceNameRules: CurationRules<String>? = null
    @SerialName("page_title") var pageTitleRules: CurationRules<String>? = null
    @SerialName("page_revision_id") var pageRevisionIdRules: ComparableCurationRules<Long>? = null
    @SerialName("page_wikidata_qid") var pageWikidataQidRules: CurationRules<String>? = null
    @SerialName("page_content_language") var pageContentLanguageRules: CurationRules<String>? = null
    @SerialName("performer_id") var performerIdRules: ComparableCurationRules<Int>? = null
    @SerialName("performer_name") var performerNameRules: CurationRules<String>? = null
    @SerialName("performer_session_id") var performerSessionIdRules: CurationRules<String>? = null
    @SerialName("performer_pageview_id") var performerPageviewIdRules: CurationRules<String>? = null
    @SerialName("performer_groups") var performerGroupsRules: CollectionCurationRules<String>? = null
    @SerialName("performer_is_logged_in") var performerIsLoggedInRules: CurationRules<Boolean>? = null
    @SerialName("performer_is_temp") var performerIsTempRules: CurationRules<Boolean>? = null
    @SerialName("performer_registration_dt") var performerRegistrationDtRules: ComparableCurationRules<Instant?>? = null

    @SerialName("performer_language_groups")
    var performerLanguageGroupsRules: CurationRules<String?>? = null

    @SerialName("performer_language_primary")
    var performerLanguagePrimaryRules: CurationRules<String?>? = null

    override fun test(event: EventProcessed): Boolean {
        return applyAgentRules(event.agentData)
                && applyMediaWikiRules(event.mediawikiData)
                && applyPageRules(event.pageData)
                && applyPerformerRules(event.performerData)
    }

    private fun applyAgentRules(data: AgentData): Boolean {
        return applyPredicate(this.agentAppInstallIdRules, data.appInstallId)
                && applyPredicate(this.agentClientPlatformRules, data.clientPlatform)
                && applyPredicate(
            this.agentClientPlatformFamilyRules,
            data.clientPlatformFamily
        )
    }

    private fun applyMediaWikiRules(data: MediawikiData): Boolean {
        return applyPredicate(this.mediawikiDatabase, data.database)
    }

    private fun applyPageRules(data: PageData): Boolean {
        return applyPredicate(this.pageIdRules, data.id)
                && applyPredicate(this.pageNamespaceIdRules, data.namespaceId)
                && applyPredicate(this.pageNamespaceNameRules, data.namespaceName)
                && applyPredicate(this.pageTitleRules, data.title)
                && applyPredicate(this.pageRevisionIdRules, data.revisionId)
                && applyPredicate(this.pageWikidataQidRules, data.wikidataItemQid)
                && applyPredicate(this.pageContentLanguageRules, data.contentLanguage)
    }

    private fun applyPerformerRules(data: PerformerData): Boolean {
        return applyPredicate(this.performerIdRules, data.id)
                && applyPredicate(this.performerNameRules, data.name)
                && applyPredicate(this.performerSessionIdRules, data.sessionId)
                && applyPredicate(this.performerPageviewIdRules, data.pageviewId)
                && applyPredicate(this.performerGroupsRules, data.groups)
                && applyPredicate(this.performerIsLoggedInRules, data.isLoggedIn)
                && applyPredicate(this.performerIsTempRules, data.isTemp)
                && applyPredicate(this.performerRegistrationDtRules, data.registrationDt)
                && applyPredicate(this.performerLanguageGroupsRules, data.languageGroups)
                && applyPredicate(this.performerLanguagePrimaryRules, data.languagePrimary)
    }

    companion object {
        private fun <T> applyPredicate(rules: Predicate<T>?, value: T?): Boolean {
            if (rules == null) return true
            if (value == null) return false
            return rules.test(value)
        }
    }
}
